includes/wikia/api/CrossWikiArticlesApiController.class.php

```diff
@@ -58,48 +58,29 @@ public function getDetails() {
 	 *
 	 * @param int $wikiId
 	 * @param int[] $articles
+	 * @param $wikiToArticleMap
 	 * @return array
 	 * @throws Exception
 	 */
-	protected function getDetailsForWiki( int $wikiId, array $articles ) {
-		$dbr = wfGetDB( DB_SLAVE, [], WikiFactory::IDtoDB( $wikiId ) );
+	protected function getDetailsForWiki( int $wikiId, array $articles, array $wikiToArticleMap ) {
+		$dbname = WikiFactory::IDtoDB($wikiId);
+		$dbr = wfGetDB( DB_SLAVE, [], $dbname );
 
 		$res = $dbr->select(
-			['page', 'page_wikia_props'],
-			'page.page_id, page_title, page_namespace, props',
+			['page'],
+			'page.page_id, page_title, page_namespace',
 			['page.page_id' => $articles],
 			__METHOD__,
 			[],
-			[
-				'page_wikia_props' => [
-					'LEFT JOIN', [
-						'page.page_id=page_wikia_props.page_id',
-						'propname' => WPP_IMAGE_SERVING
-					]
-				],
-			]
+			[]
 		);
 
 		$videos = $this->getVideosIds( $wikiId );
 		$wikiName = WikiFactory::getVarValueByName('wgSitename', $wikiId);
+		$wikiDetails = $this->getThumbnails( $dbname, $wikiToArticleMap[$wikiId] );
```

Comment:
Then through ArticlesApiController we jump to get the thumbnails data we want, but there are a lot calls/other stuff that is going on.

https://github.com/Wikia/app/blob/a3291da9dec6988c205a286b30cbb6ca4cd0c7b2/includes/wikia/api/ArticlesApiController.class.php#L796-L798
In the end we ask ImageServing class for the best image.

Reply:
I am not sure how ImageServing works, but couldn't we get an instance/recreate it per wiki in CrossWikiArticlesApiController directly to get data and skip some hoops?

So we would add:
```
	protected function getImageServing( $ids, $width, $height ) {
		return new ImageServing( $ids, $width, $height , $db?);
	}
```
I see that ImageServing takes an optional $db argument, and we already know the db that the wiki data is on - maybe we could use that?

And we could do:
```
protected function getThumbnails( $dbname, $articleIds ) {
			$is = $this->getImageServing( $ids, $width=CrossWikiArticlesApiController::THUMBNAIL_SIZE, $height=CrossWikiArticlesApiController::THUMBNAIL_SIZE, $dbname);
			$images = $is->getImages( 1 );
                          ..... some logic

```

Reply:
@MikolajBalcerek `ApiService` uses MW's default HTTP client under the hood, that in turn uses the HTTP proxy configured in the `$wgHTTPProxy` variable. In LocalSettings.php, we set this proxy to use the k8s deployment name if we detect that we are running in k8s, so any requests made through MW's HTTP abstraction layer should not leave the k8s cluster.

As for `ImageServing`, unfortunately it cannot be used in a foreign wiki context, as internally it calls code such as `Title` that will always run database queries in the context of the current wiki. An alternative solution would be to explore what we discussed some time ago, which is to send separate requests to each wiki that we'd like to serve recommendations for, instead of using the community wiki API. This would probably be more performant than the solution proposed in this PR, since this PR would use 1 request to community wiki API that would then make 1 request internally for each wiki that we need to load recommendations for. Also, on the service level, you can also execute those requests in parallel, reducing overall wait time.

@mikulskibartosz Unfortunately such controller classes do not lend themselves well to unit testing due to their lack of support for DI. Something you could try is to factor out most business logic into a separate class, and have the controller instantiate this class and pass in any required dependencies as constructor or method parameters. This would result in only  minimal logic on the controller level and could allow for unit testing the newly created class.